netty/src/main/java/io/grpc/transport/netty/NettyClientHandler.java

```diff
@@ -280,10 +283,29 @@ protected boolean isGracefulShutdownComplete() {
    * the creation request is queued.
    */
   private void createStream(CreateStreamCommand command, final ChannelPromise promise) {
-    final int streamId = getAndIncrementNextStreamId();
+    // Get the stream ID for the new stream.
+    final int streamId;
+    try {
+      streamId = getAndIncrementNextStreamId();
+    } catch (StatusException e) {
+      // Stream IDs have been exhausted for this connection. Fail the promise immediately.
+      promise.setFailure(e);
+
+      // If we haven't already started a graceful shutdown of the connection, do so now.
+      if (ctx.channel().isActive() && !connection().goAwaySent()) {
+        logger.fine("Stream IDs have been exhausted for this connection. "
+                + "Initiating graceful shutdown of the connection.");
+        try {
+          super.close(ctx, ctx.newPromise());
+        } catch (Exception ce) {
```

Comment:
Should we just add `throws Exception` to `createStream()` and `write()`? That seems more appropriate.